AAInstrumentation/optimization.py

```diff
@@ -0,0 +1,102 @@
+from __future__ import annotations
+from core import AAInstrumentationDriver, register_arguments
+from multiprocessing import Pool
+import os
+from pathlib import Path
+import subprocess
+import hashlib
+import argparse
+import sys
+from dataclasses import dataclass
+from abc import abstractmethod
+
+
+@dataclass
+class OptimizerDriver(AAInstrumentationDriver):
+    optimizers: list[Optimizer] | None = None
+
+    def __post_init__(self):
+        optimizer_list = []
+        for optimizer in self.optimizers:
+            optimizer_list.append(
+                optimizer(
+                    self.instr_path,
+                    self.exec_root,
+                    self.specbuild_dir,
+                    self.benchmark,
+                    self.initial_dir,
+                    self.instr_dir,
+                    self.groundtruth_dir,
+                    self.opt_flag,
+                    self.proc_count,
+                )
+            )
+        self.optimizers = optimizer_list
+
+    def run(self):
+        self.generate_baseline()
+        files = self.get_baseline_files()
+
+        candidates_per_file = self.get_candidates_per_file(files)
+
+        precise_sizes = {}
+        for file_ in files:
+            precise_sizes[file_] = self.run_assemble_and_measure_file(file_, 0, [])
+
+        for optimizer in self.optimizers:
+            for file_ in files:
+                print(f"Optimizing {file_} with {optimizer.description}")
+                print(f"Number of candidates: {candidates_per_file[file_]}")
+                current_minimum = optimizer.optimize(file_, candidates_per_file[file_])
+                print(f"Minimum size: {current_minimum}")
+                print(f"Compared to precise size: {precise_sizes[file_]}")
+
+
+class Optimizer(AAInstrumentationDriver):
+    @abstractmethod
+    def optimize(self, file_name: Path, num_candidates: int) -> int:
+        """Applies the optimization to the file_name, and returns minimumm size."""
+        pass
+
+
+class ImpreciseOptimizer(Optimizer):
+    """
+    This class implements the imprecise search algorithm. It is used as a
+    baseline for the other optimizers.
+    """
+
+    description = "Imprecise search"
+
+    def optimize(self, file_name: Path, num_candidates: int) -> int:
+        return self.run_assemble_and_measure_file(
+            file_name, 0, list(range(num_candidates))
+        )
```

Comment:
Does this really do anything? Other than just checking one query sequence?

Reply:
This is an implementation of one of the baselines. If you remember, next to comparing to precise AAInfo, I also compared to fully imprecise info. This instrumentation for the fully imprecise thing is literally just adding all indexes, so I decided to use it as the proof of concept for this first "optimizer".

Reply:
Good point, changed :+1: